source/controllers/sudokuWorld.d

```diff
@@ -0,0 +1,161 @@
+		callbacks();
+	}
+
+
+	private void callbacks()
+	{
+		trace("Initializing SudokuWorld callbacks...");
+		app.setOnBtnCreateClicked(&onBtnCreateClicked);
+		app.setOnBtnExitClicked(&onBtnExitClicked);
+		app.setOnButtonPress(&onButtonPress);
+	}
+
+
+	/** This is called when `btnCreate` is clicked in the main menu
+	 *
+	 * Makes the create menu content `Stack` visible
+	 */
+	private void onBtnCreateClicked(Button)
+	{
+		app.showCreateMenu();
+	}
+
+
+	/** Exit the program
+	 */
+	private void onBtnExitClicked(Button)
+	{
+		app.quit();
+	}
+
+
+	// TODO: ui:sudokuWorld: implement Play Menu
+	private void onBtnPlayClicked(Button)
+	{
+		// do something
+	}
+
+
+	// TODO: ui:sudokuWorld: implement Settings Menu
+	private void onBtnSettingsClicked(Button)
+	{
+		// do something
+	}
+
+
+	private bool onButtonPress(GdkEventButton* ev, Widget widget)
+	{
+		if (ev.button != 1)
+			return false;
+
+
+		// clear focus in the active grid
+		GridUI gridUI;
+		if (GameState.Play in sudokuControllers)
+		{
+			gridUI = sudokuControllers[GameState.Play].gridUI;
+		}
+		else if (GameState.Solution in sudokuControllers)
+		{
+			gridUI = sudokuControllers[GameState.Solution].gridUI;
+			_createMenuController.setCellInfoText(0,0,0);
+		}
+		else if (GameState.Create in sudokuControllers)
+		{
+			gridUI = sudokuControllers[GameState.Create].gridUI;
+			_createMenuController.setCellInfoText(0,0,0);
+		}
+		gridUI.cleanFocus();
+		return false;
+	}
+
+
+
+// Getters/Setters
+
+	public void addSudokuController(SudokuController sudokuController)
+	{
+		sudokuControllers[sudokuController.gameState] = sudokuController;
+	}
+
+
+	public auto addSudokuController(GameState state, SudokuType type)
+	{
+		Grid grid = new Grid(Sudoku.dimension(type).expand);
+		grid.initialize(new int[][](grid.height, grid.width));
+		GridUI gridUI = new GridUI(type);
+		SudokuController sd = new SudokuController(grid, gridUI, state, this);
+		return sudokuControllers[state] = sd;
+	}
+
+
+	public auto addSudokuController(GameState state, SudokuController sudokuController)
+	{
+		SudokuType type = sudokuController.gridUI.type;
+		Grid grid = new Grid(Sudoku.dimension(type).expand);
+		int[][] digits = sudokuController.grid.toDigit();
+		grid.initialize(digits.dup);
+		GridUI gridUI = new GridUI(type);
+		SudokuController sd = new SudokuController(grid, gridUI, state, this);
+		return sudokuControllers[state] = sd;
+	}
+
+
+	public void deleteSudokuController(GameState state)
+	{
+		SudokuController* p = (state in sudokuControllers);
+		if (p !is null)
+		{
+			SudokuController sc = *p;
+			sc.gridUI.destroy();
+			sudokuControllers.remove(state);
+		}
+	}
+
+
+	public auto getSudokuController(GameState state)
+	{
+		return sudokuControllers[state];
+	}
+
+
+	public auto createMenuController() @property
+	{
+		return _createMenuController;
+	}
+
+
+	public auto app() @property
+	{
+		return _app;
+	}
+
+
+	private CreateMenuController _createMenuController;
```

Comment:
The `@property` functions existent are getters. Other classes might need those variables but shouldn't be able to change them. Actually, right now, this isn't completely correct because the return value isn't `const`. I'll change it.

Regarding the use of `_`, is part of the Dlang style: https://dlang.org/dstyle.html#naming_conventions

Reply:
Ah, my bad, I thought you did it differently from the code style. But in that case, you should change every private field to have `_`, for consistency.

Regarding the getters, I just don't see much of the point of encapsulating them when you're returning the same object. The getters here return a direct reference to the object, so it's being modified the same way as accessing it directly.

Reply:
Not really. You can specify you want to pass the object as `const` which means you cannot change the object or any data in it. You cannot access any function not declared as `const' as well. So this is usefull is you want another instance who needs some value of it, but doesn't have permissions to change it.

Reply:
Ahh, okay, in that case, a getter would make more sense then.